libp2p/network/url_conversion_test.go

```diff
@@ -0,0 +1,159 @@
+			require.NoError(t, err)
+			require.NotNil(t, actualMultiaddr)
+			require.Regexp(t, regexp.MustCompile(testCase.expetedMultiaddrRegex), actualMultiaddr.String())
+		})
+	}
+}
+
+const (
+	FQDN = "fqdn"
+	IP4  = "ip4"
+	IP6  = "ip6"
+)
+
+func TestPeerMultiAddrFromServiceURL_error(t *testing.T) {
+	hostnames := map[string]string{
+		FQDN: "www.google.com",
+		IP4:  "142.250.181.196",
+		IP6:  "2a00:1450:4005:802::2004",
+	}
+
+	testCases := []struct {
+		name             string
+		serviceUrlFormat string
+		// TECHDEBT: assert specific errors.
+		expectedErrContains string
+	}{
+		// Usage of scheme is invalid.
+		{
+			"FQDN with scheme",
+			"tcp://%s:8080",
+			"no such host",
+		},
+
+		// Port **number** is required
+		{
+			"invalid port number",
+			"%s:abc",
+			"invalid port",
+		},
+		{
+			"missing port number",
+			"%s:",
+			"unexpected end of multiaddr",
+		},
+		// TODO: this case is tricky to detect as IPv6 addresses
+		// can omit hextet delimiters and still be valid.
+		// (see: Address_representation)
+		// {
+		// 	"missing port number and delimiter",
+		// 	"%s",
+		// },
+	}
+
+	for _, testCase := range testCases {
+		for hostType, hostname := range hostnames {
+			testName := fmt.Sprintf("%s/%s", testCase.name, hostType)
+			t.Run(testName, func(t *testing.T) {
+				serviceURL := fmt.Sprintf(testCase.serviceUrlFormat, hostname)
+				actualMultiaddr, err := Libp2pMultiaddrFromServiceUrl(serviceURL)
+				// TECHDEBT: assert specific errors
+				// Print resulting multiaddr to understand why no error.
+				require.ErrorContainsf(t, err, testCase.expectedErrContains, fmt.Sprintf("actualMultiaddr: %s", actualMultiaddr))
+			})
+		}
+	}
+}
+
+func TestServiceUrlFromLibp2pMultiaddr_success(t *testing.T) {
+	testCases := []struct {
+		name         string
+		multiaddrStr string
+		expectedUrl  string
+	}{
+		{
+			"IPv4",
+			"/ip4/142.250.181.196/tcp/8080",
+			"142.250.181.196:8080",
+		},
+		{
+			"IPv6",
+			"/ip6/2a00:1450:4005:802::2004/tcp/8080",
+			"2a00:1450:4005:802::2004:8080",
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			addr, err := multiaddr.NewMultiaddr(testCase.multiaddrStr)
+			require.NoError(t, err)
+
+			actualUrl, err := ServiceUrlFromLibp2pMultiaddr(addr)
+			require.NoError(t, err)
+			require.NotNil(t, actualUrl)
+			require.Equal(t, testCase.expectedUrl, actualUrl)
+		})
+	}
+}
+
+func TestServiceUrlFromLibp2pMultiaddr_error(t *testing.T) {
+	testCases := []struct {
+		name                string
+		multiaddrStr        string
+		expectedErrContains string
+	}{
+		{
+			"FQDN",
+			"/dns/www.google.com/tcp/8080",
+			"unsupported network protocol",
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			addr, err := multiaddr.NewMultiaddr(testCase.multiaddrStr)
+			require.NoError(t, err)
+
+			_, err = ServiceUrlFromLibp2pMultiaddr(addr)
+			// DISCUSS: asserting specific errors instead.
```

Comment:
I think this is reasonable. Did you want to discuss if we should do it everywhere?

Reply:
Yes, I think it would be good to discuss it. I'm still developing my own opinion but at the moment I would prefer not to necessarily encourage emulating this. 

I think before we put too much effort into action, it would be prudent to nail down some reasonable requirements / acceptance criteria for errors / error handling conventions. I've started issue #519 for this purpose. In absence of such requirements, the intention behind wrapping errors like I've been doing here is mainly to aid in debugging.

Reply:
Dave Cheney (in a [GopherCon 2016 presentation](https://youtu.be/lsBF58Q-DnY?t=701)) talks about preferring to assert for errors by behavior over type which mitigates problems resulting from other solutions. So perhaps, asserting for specific errors isn't quite the right direction.

Reply:
> I've started issue #519 for this purpose. In absence of such requirements, the intention behind wrapping errors like I've been doing here is mainly to aid in debugging.

Noted. Seems like there's nothing actionable to discuss in this commit, but will keep it in mind.